### PR TITLE
`o3.Norm`/`NormActivation`: specialized code and gradient fix

### PR DESCRIPTION
**Pull request by Linux-cpp-lisp**
## Description
- Specialized code for `o3.Norm` that avoids both einsums and concatenations (uses `segment_csr` for fast sums)
- `epsilon` option in `o3.Norm` for avoiding gradient NaNs when feature has norm zero
- Use the above `epsilon` in `NormActivation` to avoid gradient NaNs
- Option to return squared norms in `o3.Norm`.

## Motivation and Context
This updated option now correctly reproduces `norm_with_epsilon` from TFN (https://github.com/tensorfieldnetworks/tensorfieldnetworks/blob/master/tensorfieldnetworks/utils.py#L22) and avoids, to the extend possible with norm nonlinearities, the gradient/asymptote issues at zero.

## How Has This Been Tested?

- e3nn test suite
- new tests
- resolved issue in real world network

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [X] I have read the [**CONTRIBUTING**](https://github.com/e3nn/e3nn/blob/main/CONTRIBUTING.md) document.
- [X] My code follows the code style of this project.
- [X] I have updated the documentation (if relevant).
- [X] I have added tests that cover my changes (if relevant).
- [X] All new and existing tests passed.
- [X] I have updated the [Changelog](https://github.com/e3nn/e3nn/blob/main/ChangeLog.md).

### CODE REVIEW

**Comment by Linux-cpp-lisp** [comment]
Benchmarking with this script:
```python
import torch
from norm_test import SlowNorm
from e3nn.o3 import Norm
from torch.utils.benchmark import Timer
ir = "16x0o + 16x0e + 16x1o + 16x1e + 16x2o + 16x2e"
l_slow = SlowNorm(ir)
l = Norm(ir)
n = 2000
inps = [l.irreps_in.randn(200, -1) for _ in range(n + 100)]
inps_iter_fast = iter(inps)
inps_iter_slow = iter(inps)
t_fast = Timer(stmt="out = l(next(inps_iter_fast))", globals={'l': l, 'inps_iter_fast': inps_iter_fast})
print(t_fast.timeit(n))
t_slow = Timer(stmt="out = l_slow(next(inps_iter_slow))", globals={'l_slow': l_slow, 'inps_iter_slow': inps_iter_slow})
print(t_slow.timeit(n))
```
revealed that the `segment_csr` specialized code was only faster for small batch sizes (<100). And while the speedups for very small batch sizes (~10) were significant, for batch sizes larger than 100 it became significantly slower.

I've taken the bugfixes/features from this and put them in #258 without the optimization, and am closing this PR and leaving it for reference.
